### ISSUE TITLE
idea: add self-contained ipopt 

### ISSUE THREAD

**Issue by yang-sjtu** [conda-forge/dolfin-adjoint-feedstock#6]
Dear maintainer,

I'm recently checking the possibility to move from docker to conda to continue my work on dolfin-adjoint, what I suffer is that both fenics and dolfin-adjoint working properly in conda env, but ipopt solver becomes an issue. I tried to install ipopt seperately from conda-forge but still have the trouble to import the ipopt solver. It has been noticed that moola is self-contained in the conda version dolfin-adjoint, so i'm wandering is it possible to add self-contained ipopt solver like what has been provided in the dolfin-adjoint docker image (despite of compiling the linear solver from HSL). I'm not sure if there is any practical reason that ipopt is not self-contained yet, but if possible, it would be nice that the conda version dolfin-adjoint is consistent with the docker verison.

Best regards,
Yang

**Comment by mariusvniekerk** [comment]
Hey @yang-sjtu feel free to make a pr to this feedstock if needed

**Comment by minrk** [comment]
I don't believe any changes are necessary here to enable ipopt, since pyadjoint does not link ipopt directly, it loads it at runtime via [pyipopt](https://github.com/xuy/pyipopt). So all that's needed is a pyipopt package for conda-forge, which will be hard to do at the moment, since there has not been a published release that I can find (though there are version numbers in the changelog).

This patch allowed me to build xuy/pyipopt@f93dbdc in a conda env:

```diff
diff --git a/setup.py b/setup.py
index 5899b61..1ec2fee 100644
--- a/setup.py
+++ b/setup.py
@@ -12,7 +12,8 @@
 # for my linux distribution was buggy,
 # so by the time you read this the bugs have probably been fixed
 # and you will want to specify a different directory here.
-IPOPT_DIR = '/usr/local/'
+import sys
+IPOPT_DIR = sys.prefix
 
 import os
 from distutils.core import setup
@@ -48,11 +49,7 @@ pyipopt_extension = Extension(
         #extra_link_args=['-Wl,--rpath','-Wl,'+ IPOPT_LIB],
         library_dirs=[IPOPT_LIB],
         libraries=[
-            'ipopt', 'coinblas',
-            #'coinhsl',
-            'coinmumps',
-            'coinmetis',
-            'coinlapack','dl','m',
+            'ipopt',
             ],
         include_dirs=[numpy_include, IPOPT_INC],
         )
```

The process being:

```bash
conda create -n adjoint dolfin-adjoint ipopt
conda activate adjoint
git clone https://github.com/xuy/pyipopt
# apply patch above
pip install ./pyipopt
```

Then you have dolfin-adjoint with ipopt support.

If there is a change to make on this repo, it would be to add pyipopt as a dependency of dolfin-adjoint once it's released and then packaged on conda-forge, but it can be used today.
